### PR TITLE
#3196 Add a facility to add a connect in a Dyd file

### PR DESCRIPTION
**Pull request by barondim**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [x] the corresponding milestone was added in the ticket and in this PR
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated

### CODE REVIEW

File: util/updateXML/content/sources/Dyd/Connects.py [hunk 1]
```diff
@@ -75,6 +76,52 @@ def change_var_name(self, current_var, new_var):
...
+                raise UnknownDydElementError(self.__parent_xml_tree.tag)
+
+        # check the connect we want to create doesn't exist
+        connect_xpath = './dyn:connect[@id1="' + self.__model_id  + '" and @var1="' + var1 + '" and @id2="' + id2 + '" and @var2="' + var2 + '"]'
+        connect = self.__parent_xml_tree.xpath(connect_xpath, namespaces=NAMESPACE_URI)
+        if len(connect) != 0:
+            raise ConnectAlreadyExists(self.__model_id, var1, id2, var2, self.__parent_xml_tree.base)
+
+        # add connect in dyd file
+        if self.__connect_type == ConnectType.initConnect:
```
Comment (rosiereflo) [inline util/updateXML/content/sources/Dyd/Connects.py:114]: WHere does this information comes from?

File: util/updateXML/content/sources/Dyd/Connects.py [hunk 2]
```diff
@@ -75,6 +76,52 @@ def change_var_name(self, current_var, new_var):
...
+                if len(connected_udm) == 0:
+                    raise ModelNotFoundError(id2, self.__parent_xml_tree.base)
+            else:
+                raise UnknownDydElementError(self.__parent_xml_tree.tag)
+
+        # check the connect we want to create doesn't exist
+        connect_xpath = './dyn:connect[@id1="' + self.__model_id  + '" and @var1="' + var1 + '" and @id2="' + id2 + '" and @var2="' + var2 + '"]'
+        connect = self.__parent_xml_tree.xpath(connect_xpath, namespaces=NAMESPACE_URI)
+        if len(connect) != 0:
+            raise ConnectAlreadyExists(self.__model_id, var1, id2, var2, self.__parent_xml_tree.base)
```
Comment (rosiereflo) [inline util/updateXML/content/sources/Dyd/Connects.py:111]: It means that we cannot apply this to a testcase already migrated which is an issue. In this case we should just ignore the connect.

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: util/updateXML/content/sources/Dyd/Connects.py (hunk 1 above)
Comment (barondim) [reply util/updateXML/content/sources/Dyd/Connects.py:122]: A class representing an XML model can have an **init_connects** attribute (to manipulate dyn:initConnect) and a **connects** attribute (to manipulate dyn:connect). The **init_connects** attribute and **connects** attribute are both instances of Connects class.
The **init_connects** and **connects** attribute have both an attribute **self.__connect_type** when they are constructed to know if they manipulate dyn:initConnect or dyn:connect

Comment (rosiereflo) [reply util/updateXML/content/sources/Dyd/Connects.py:122]: No I mean we are creating a connect from scratch here, where does the user says that he wants to create a initConnect or a connect?

Comment (barondim) [reply util/updateXML/content/sources/Dyd/Connects.py:108]: If the user uses the **init_connect** attribute, he will create a dyn:initConnect.
If the user uses the **connect** attribute, he will create a dyn:connect

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**